fix(utxo-lib): fix (de)serialization methods in UtxoPsbt

davidkaplanbitgo opened the pull request:
There are a few things that are fixed in this PR to ensure proper serialization and deserialization of PSBTs:
 * Format BIP32 Derivation path so that it can be deserialized with a relative path
 * Changes he `UtxoPsbt.fromBuffer` method to use `this.fromHex` instead of `UtxoPsbt.fromHex` so that it returns the right class (`DashPsbt`, `ZcashPsbt`, or `UtxoPsbt`)

Associated previous PRs:
* https://github.com/BitGo/bitcoinjs-lib/pull/54
* https://github.com/BitGo/bip174/pull/1

[BG-54756](https://bitgoinc.atlassian.net/browse/BG-54756)

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

[BG-54756]: https://bitgoinc.atlassian.net/browse/BG-54756?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

brandonblack commented on modules/utxo-lib/src/bitgo/UtxoPsbt.ts:
```diff
@@ -145,7 +145,21 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
+    });
+
+    // Remove unknownKeyVals from ins and outs if they are empty.
+    // We do this because we want to have the psbt be changeless from the transaction it is wrapping.
```
Not sure I understand this reasoning. An empty list-map is equivalent to it not existing in the PSBT serialization format.

davidkaplanbitgo replied:
It's true - but when you deserialize a psbt hex, these empty arrays do not show up in the `unspents` and `outputs`. If you were to do a comparison of 

```typescript
...
const psbtHex = originalPsbt.toHex();
const deserializedPsbt = UtxoPsbt.fromHex(psbtHex, { network });
assert.deepStrictEqual(deserializedPsbt, originalPsbt)
```
This would fail because of these empty `unknownKeyVals`

brandonblack replied:
Hmm, that seems like something to fix in test code rather than here in production code then.

brandonblack replied:
Or something to fix in a better written comparison (for example, one way to compare PSBTs is to serialize them to a buffer and compare the buffers, since _that_ is a canonical representation).

OttoAllmendinger commented on modules/utxo-lib/src/bitgo/wallet/chains.ts:
```diff
@@ -110,3 +110,11 @@ export function isSegwit(v: ChainCode): boolean {
   return segwitCodes.includes(v);
 }
+
+/**
+ * Prepend bip32 derivation path with 'm/' if it is not there
+ * @param path
+ */
+export function formatCorrectBip32DerivationPath(path: string): string {
+  return path.slice(0, 2) !== 'm/' ? 'm/' + path : path;
+}
```
as discussed on slack - paths that don't start with `m/` must be supported, just adding `m/` is not a solution here

davidkaplanbitgo replied:
Looking upstream now

davidkaplanbitgo committed:
fix(utxo-lib): fix UtxoPsbt (de)serialization

Fix (tap)bip32Derivation path deserialization by passing in optional
flag to make paths absolute. Additionally, fix psbt deserialization
methods so that child classes inherit them correctly

BG-54756

brandonblack commented on modules/utxo-lib/src/bitgo/transaction.ts:
```diff
@@ -37,25 +37,37 @@ export function createTransactionFromBuffer<TNumber extends number | bigint = nu
+export function createPsbtFromBuffer(
+  buf: Buffer,
+  network: Network,
+  bip32PathsAbsolute = true
```
This is a BitGo library, I think we should make this false here.

brandonblack reviewed (changes requested):
Couple things to improve, nice work on getting the @bitgo-forks setup and sorting through all this dependency fork blah blah blah to get this fixed :)

davidkaplanbitgo committed:
chore(utxo-lib): bump and consolidate bitcoinjs-lib dependencies

BG-54756

davidkaplanbitgo commented:
Should be good now.

brandonblack commented on modules/utxo-lib/src/bitgo/transaction.ts:
```diff
@@ -37,25 +37,37 @@ export function createTransactionFromBuffer<TNumber extends number | bigint = nu
+export function createPsbtFromHex(
+  hex: string,
+  network: Network,
+  bip32PathsAbsolute = true
```
default false here too?

brandonblack reviewed (changes requested):
1 nit

davidkaplanbitgo commented:
NOW it should be good lol

davidkaplanbitgo commented on modules/utxo-lib/src/bitgo/zcash/ZcashPsbt.ts:
```diff
@@ -128,6 +124,9 @@ export class ZcashPsbt extends UtxoPsbt<ZcashTransaction<bigint>> {
     this.tx.consensusBranchId = getDefaultConsensusBranchIdForVersion(network, version);
   }
 
+  // For Zcash transactions, we do not have to have non-witness utxo data for non-segwit
+  // transactions because zcash hashes the value directly and therefore does not need to
+  // check the previous transaction.
```
Realized [this comment](https://github.com/BitGo/BitGoJS/pull/3210/files#r1087261135) after merging the other PR, adding the comment here
